matcher/matcher.go

```diff
@@ -7,52 +7,191 @@ import (
+		return func(_ []byte) bool { return true }
+	case 1:
+		// when there is only one element in matcherFuncs we can directly return
+		// it and save the overhead of having to wrap another function around it
+		return m[0]
+	default:
+		return func(s []byte) bool {
+			for _, matcherFunc := range m {
+				if !matcherFunc(s) {
+					return false
+				}
+			}
+			return true
+		}
+	}
+}
+
+func New(prefix, notPrefix, sub, notSub, regex, notRegex string) (Matcher, error) {
+	match := Matcher{
+		Prefix:    prefix,
+		NotPrefix: notPrefix,
+		Sub:       sub,
+		NotSub:    notSub,
+		Regex:     regex,
+		NotRegex:  notRegex,
+	}
 	err := match.updateInternals()
 	if err != nil {
-		return nil, err
+		return match, err
 	}
 	return match, nil
 }
 
 func (m *Matcher) String() string {
-	return fmt.Sprintf("<Matcher. prefix:%q, sub: %q, regex: %q>", m.Prefix, m.Sub, m.Regex)
+	return fmt.Sprintf("<Matcher. prefix:%q, notPrefix:%q, sub: %q, notSub: %q, regex: %q, notRegex:%q>", m.Prefix, m.NotPrefix, m.Sub, m.NotSub, m.Regex, m.NotRegex)
 }
 
+// updateInternals checks which filters this matcher should use and generates a
+// filter function for each of them.
+// Then it combines all these filter functions into one function
+// which it assigns to Matcher.Match.
+// Furthermore, it combines all filter functions except the "regex" filter into
+// another single filter function and assigns it to Matcher.MatchAllExceptRegex.
 func (m *Matcher) updateInternals() error {
 	m.prefix = []byte(m.Prefix)
-	m.substring = []byte(m.Sub)
+	m.notPrefix = []byte(m.NotPrefix)
+	m.sub = []byte(m.Sub)
+	m.notSub = []byte(m.NotSub)
 	if len(m.Regex) > 0 {
 		regexObj, err := regexp.Compile(m.Regex)
 		if err != nil {
 			return err
 		}
 		m.regex = regexObj
+		m.prefixFromRegex = regexToPrefix(m.Regex)
 	}
+	if len(m.NotRegex) > 0 {
+		regexObj, err := regexp.Compile(m.NotRegex)
+		if err != nil {
+			return err
+		}
+		m.notRegex = regexObj
+	}
+
+	var usedFuncs matcherFuncs
+	var usedFuncsExceptRegex matcherFuncs
+	if len(m.prefix) > 0 {
+		usedFuncs = append(usedFuncs, m.matchPrefix)
+		usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchPrefix)
+	}
+	if len(m.notPrefix) > 0 {
+		usedFuncs = append(usedFuncs, m.matchNotPrefix)
+		usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchNotPrefix)
+	}
+	if len(m.sub) > 0 {
+		usedFuncs = append(usedFuncs, m.matchSub)
+		usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchSub)
+	}
+	if len(m.notSub) > 0 {
+		usedFuncs = append(usedFuncs, m.matchNotSub)
+		usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchNotSub)
+	}
+	if m.regex != nil {
+		usedFuncs = append(usedFuncs, m.matchRegex)
+		if len(m.prefixFromRegex) > 0 {
+			usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchPrefixFromRegex)
+		}
+	}
+	if m.notRegex != nil {
+		usedFuncs = append(usedFuncs, m.matchNotRegex)
+		usedFuncsExceptRegex = append(usedFuncsExceptRegex, m.matchNotRegex)
```

Comment:
why are we appending the regex matcher to usedFuncs**exceptregex** ?
also some similar cases above.

Reply:
For the positive match it's only doing it if there is a fixed prefix, we should be able to do the same here but with a match that only fires if the fixed prefix doesn't match

Reply:
> why are we appending the regex matcher to usedFuncsexceptregex ?
also some similar cases above.

Probably the naming doesn't make this clear enough. But `usedFuncsExceptRegex` should be all filters except the positive `regex` filter, so it should include the `notRegex` filter. The reason why we need `usedFuncsExceptRegex` is to be able to do the pre-matching here: https://github.com/grafana/carbon-relay-ng/pull/421/files#diff-4c012cb82fa84751e2387047fb4d835dR226

In the meantime you told me that we should focus on blocking `.AddMaybe()` for as short as possible, which makes sense. So now I'm thinking it would probably be better to remove the `notRegex` filter from `usedFuncsExceptRegex` and only do it from `.AddMaybe()` if `dropRaw=true`.

Reply:
How is that: bf16475